transport_plugins/bled112/scripts/run_mock_observer.py

```diff
@@ -0,0 +1,48 @@
+"""
+Script to test throughput of the gateway
+"""
+
+import argparse
+import time
+import signal
+import sys
+import logging
+import serial
+# TODO import properly, now one have to specify PYTHONPATH=../test
+from util.mock_adv_bled112 import MockAdvertisingBLED112
+
+logging.basicConfig(format='%(asctime)s - %(message)s', level=logging.INFO)
+
+def main():
+    argparser = argparse.ArgumentParser()
+    argparser.add_argument("--max-advertisements-per-second", type=int, required=True)
+    argparser.add_argument("--port", help="/dev/pts/X", required=True)
+    args = argparser.parse_args()
+
+    pts = serial.Serial(args.port, timeout=1)
+
+    bled112 = MockAdvertisingBLED112(3)
+
+    while not bled112.scanning:
+        data = pts.readline()
+        if data:
+            response = bled112.generate_response(data)
+            pts.write(response)
+
+    logging.info("Start sending advertisements")
+
+    while True:
+        current_time = time.time()
+
+        data = bled112.generate_multiple_adv_packets(args.max_advertisements_per_second)
```

Comment:
When I use this according to the README, I get tons of the following log message:
![image](https://user-images.githubusercontent.com/2424759/62149593-efe10300-b2b0-11e9-9792-91f71e23e4b3.png)

It seems like there's a synchronization problem or something where maybe .  

The emulator may need to be a bit more complicated with foreground thread reading packets and responding to commands and a background thread sending advertisements while scanning is enabled.  You can use the `AsyncPacketBuffer` as an example of how it could be done.

Reply:
Yes, on gateway restart there could be such a problem.
It looks like AsyncPacketBuffer resolves it.